src/dsp_tools/analyse_xml_data/extract_links_from_XML.py

```diff
@@ -23,85 +25,145 @@ def create_classes_from_root(root: etree._Element) -> list[ResptrLink] and list[
-) -> list[ResptrLink] | None and list[XMLLink] | None and list[str] | None:
+) -> tuple[list[ResptrLink] | None, list[XMLLink] | None, list[str] | None]:
     subject_id = resource.attrib.get("id")
+    if not subject_id:
+        return None, None, None
     all_used_ids = [subject_id]
     resptr_links, xml_links = _get_all_links_one_resource(resource)
+    resptr_link_objects = []
+    xml_link_objects = []
     if resptr_links:
         all_used_ids.extend(resptr_links)
-        weight_dict = _make_weighted_resptr_links(resptr_links)
-        resptr_links = [ResptrLink(subject_id=subject_id, object_id=k, edge_weight=v) for k, v in weight_dict.items()]
+        resptr_link_objects = [ResptrLink(subject_id, object_id) for object_id in resptr_links]
     if xml_links:
         all_used_ids.extend(chain.from_iterable(xml_links))
-        xml_links = [XMLLink(subject_id=subject_id, object_link_ids=x) for x in xml_links]
-    return resptr_links, xml_links, all_used_ids
+        xml_link_objects = [XMLLink(subject_id, x) for x in xml_links]
+    return resptr_link_objects, xml_link_objects, all_used_ids
 
 
-def _make_weighted_resptr_links(resptr_links: list[str]) -> dict[str, int]:
-    weight_dict = {link: 0 for link in set(resptr_links)}
-    for link in resptr_links:
-        weight_dict[link] += 1
-    return weight_dict
-
-
-def _get_all_links_one_resource(resource: etree._Element) -> list[str] | None and list[set[str]] | None:
-    resptr_links = []
-    xml_links = []
+def _get_all_links_one_resource(resource: etree._Element) -> tuple[list[str], list[set[str]]]:
+    resptr_links: list[str] = []
+    xml_links: list[set[str]] = []
     for prop in resource.getchildren():
         match prop.tag:
             case "{https://dasch.swiss/schema}resptr-prop":
-                links = _extract_id_one_resptr_prop(prop)
-                match links:
-                    case list():
-                        resptr_links.extend(links)
-                    case None:
-                        continue
+                resptr_links.extend(_extract_id_one_resptr_prop(prop))
             case "{https://dasch.swiss/schema}text-prop":
-                links = _extract_id_one_text_prop(prop)
-                match links:
-                    case list():
-                        xml_links.extend(links)
-                    case None:
-                        continue
-    if len(resptr_links) == 0:
-        resptr_links = None
-    if len(xml_links) == 0:
-        xml_links = None
+                xml_links.extend(_extract_id_one_text_prop(prop))
     return resptr_links, xml_links
 
 
 def _extract_id_one_resptr_prop(resptr_prop: etree._Element) -> list[str]:
-    return [x.text for x in resptr_prop.getchildren()]
+    return [x.text for x in resptr_prop.getchildren() if x.text]
 
 
-def _extract_id_one_text_prop(text_prop: etree._Element) -> list[set[str]] | None:
+def _extract_id_one_text_prop(text_prop: etree._Element) -> list[set[str]]:
```

Comment:
if there are no text prop there would be no sets in the lists. So should this not be a problem with the typing?

Reply:
Type hints for containers (here: the list) tell you what kind of item will be in the container. There can also be 0 of those items in the container, without any problems